File: libwebrtc-sys/src/video_decoder.cc
#include "libwebrtc-sys/src/lib.rs.h"

void ArcasDecodedImageCallback::SetCallback(webrtc::DecodedImageCallback* cb) {
    this->cb = cb;
}

int32_t ArcasDecodedImageCallback::decoded(webrtc::VideoFrame& frame) const {
    return this->cb? this->cb->Decoded(frame) : 0;
}

bool ArcasVideoDecoder::Configure(const webrtc::VideoDecoder::Settings& settings) {
    return true;
}

int32_t ArcasVideoDecoder::Decode(
    const webrtc::EncodedImage& image,
    bool missing_frames,
    int64_t render_times_ms
) {
    return api->decode(image, missing_frames, render_times_ms, cb);
}

int32_t ArcasVideoDecoder::RegisterDecodeCompleteCallback(
    webrtc::DecodedImageCallback* cb
) {
    this->cb.SetCallback(cb);
    return 0;
}

int32_t ArcasVideoDecoder::Release() {
    return api->release();
}

int ArcasVideoDecoder::GetNumFramesReceived() const {
    return api->get_num_frames_received();
}

File: libwebrtc-sys/src/video_decoder.rs
use std::pin::Pin;

pub struct DecodedImageCallback<'a> {
    cb: &'a crate::ffi::ArcasDecodedImageCallback,
}

impl<'a> DecodedImageCallback<'a> {
    pub fn new(cb: &'a crate::ffi::ArcasDecodedImageCallback) -> Self {
        Self { cb }
    }

    pub fn decoded(&self, frame: Pin<&mut crate::ffi::ArcasCxxVideoFrame>) -> i32 {
        self.cb.decoded(frame)
    }
}

pub trait VideoDecoderImpl {
    fn decode(
        &mut self,
        image: &crate::ffi::ArcasCxxEncodedImage,
        missing_frames: bool,
        render_times_ms: i64,
        cb: DecodedImageCallback,
    ) -> i32;

    fn release(&mut self) -> i32;

    fn get_num_frames_received(&self) -> i32;
}

pub struct VideoDecoderProxy {
    decoder: Box<dyn VideoDecoderImpl>,
}

impl VideoDecoderProxy {
    pub fn new(decoder: Box<dyn VideoDecoderImpl>) -> Self {
        Self { decoder }
    }

    pub fn decode(
        &mut self,
        image: &crate::ffi::ArcasCxxEncodedImage,
        missing_frames: bool,
        render_times_ms: i64,
        decoded_image_callback: &crate::ffi::ArcasDecodedImageCallback,
    ) -> i32 {
        self.decoder.decode(
            image,
            missing_frames,
            render_times_ms,
            DecodedImageCallback::new(decoded_image_callback),
        )
    }

    pub fn release(&mut self) -> i32 {
        self.decoder.release()
    }

    pub fn get_num_frames_received(&self) -> i32 {
        self.decoder.get_num_frames_received()
    }
}
